src/blcu/download.go

```diff
@@ -123,5 +130,80 @@ func (blcu *BLCU) writeDownloadFile(data downloadData) error {
 		return err
 	}
 
-	return os.WriteFile(path.Join(blcu.config.DownloadPath, data.Board+".bin"), data.Payload, 0777)
+	return os.WriteFile(path.Join(blcu.config.DownloadPath, fmt.Sprintf("%s-%d.bin", board, time.Now().Unix())), data, 0777)
+}
+
+type DownloadResult struct {
+	Data []byte
+	Err  error
+}
+
+type Download struct {
+	output         bytes.Buffer
+	Result         <-chan DownloadResult
+	resultChan     chan<- DownloadResult
+	Percentage     <-chan float64
+	percentageChan chan<- float64
+	notify         chan int
+	total          int
+	writeErr       error
+}
+
+func NewDownload(size int) Download {
+	resultChan := make(chan DownloadResult)
+	percentageChan := make(chan float64, 100)
+
+	download := Download{
+		output:         bytes.Buffer{},
+		Result:         resultChan,
+		resultChan:     resultChan,
+		Percentage:     percentageChan,
+		percentageChan: percentageChan,
+		notify:         make(chan int),
+		total:          size,
+		writeErr:       nil,
+	}
+
+	go download.consumeNotifications()
+
+	return download
+}
+
+func (download *Download) Write(p []byte) (n int, err error) {
+	if download.writeErr != nil {
+		return 0, download.writeErr
+	}
+	n, err = download.output.Write(p)
+	if err != nil {
+		download.writeErr = err
+		download.abort(err)
+	} else {
+		download.notify <- n
+	}
+	return n, err
+}
+
+func (download *Download) abort(err error) {
```

Comment:
Why do you need to send the error through a channel? receiver.WriteTo returns an error, so consumers should check that error to determine if the download failed